### PR TITLE
Rework how gratuitous advertisements (ARP/NDP) work

### LINKED ISSUE #584: L2 mode doesn't resend gratuitous arp when node is diconnected and reconnected
If the master node is disconnected for some time, another node will start to announce as the master. But after the original master node is reconnected, it doesn't resend gratuitous arp because it doesn't think it ever lose mastership. But now it's in a inconsistent state where traffic still go through the 2nd master because gratuitous arp is not sent. This inconsistent state may take a long time due to ongoing traffic with wrong MAC refreshing gateway arp table.

Shall we consider sending garp all the time instead of only the first 5 seconds?

     - Version of MetalLB v0.9.3
     - Version of Kubernetes v1.15
     - Name and version of network addon (e.g. Calico, Weave...) calico
     - Whether you've configured kube-proxy for iptables or ipvs mode iptable

### PR DESCRIPTION
**Pull request by champtar**
With our current approach, we only do gratuitous advertisements on the first
SetBalancer() call, and we don't resend any gratuitous advertisements on the next
calls to reduce the amount of "spam" in the network.
This was working pretty well when we were using K8S API to do all the decisions.
Now that we are also using MemberList status, our decisions are based on eventually consistent
information, and there are at least 2 cases where we need to resend gratuitous advertisements
even if the information that we have makes us think there were no changes in ownership:

1) Split brain with no ownership changes:
3 nodes A B C, A owns the LoadBalancer IP I, cluster is clean.
Now for some reason C can't talk to A and B anymore, and our algorithm
in ShouldAnnounce() continues to pick A as the owner of I.
As there were no changes for A, A doesn't send any gratuitous advertisement.
As C thinks it is alone, it thinks it owns I and sends gratuitous advertisements.
Some seconds later C rejoins A & B, C stops sending gratuitous advertisements,
but A continues to be the owner and doesn't send any gratuitous advertisement.
Depending on the switches' inner working, traffic might continue to go to C for a long time.

2) Race condition on ForceSync:
3 nodes A B C, A owns the LoadBalancer IP I, cluster is clean.
A becomes really slow (cpu limits or ...) and memberlist on B and C decides that A is not part of the memberlist cluster
anymore. B and C each start a ForceSync(), one of B or C becomes the owner of I and starts gratuitous advertisements for I.
A starts to respond again to memberlist and rejoins the cluster, while doing its first ForceSync().
A thinks it was always the owner of I and doesn't send any gratuitous advertisement.

The idea of this patch is to send gratuitous advertissements for 5 seconds from the last SetBalancer() call,
instead of the last time we think we became the owner.
To ensure there is only 1 sender for each IP, we use only one goroutine for all gratuitous advertisement calls.
As gratuitous() was using Lock() (ie exclusive lock), we were sending at most 1 gratuitous advertisement at a time,
so we know that this is fine performance wise, but it might be burstier than before.

This fixes #584

### CODE REVIEW

**Comment by champtar** [comment]
@russellb if you want to have a look

**Comment by russellb** [comment]
> @russellb if you want to have a look

thanks!  Can you explain a bit more in the description how this fixes #584?  The current description and commit message just talks about the lock change.

**Comment by champtar** [comment]
@russellb added way more explanations, tell me if this is enough

File: internal/layer2/announcer.go
Comment (russellb) [inline internal/layer2/announcer.go:229]: It looks like these changes make it so we're sending GARP for each service using the same IP, bypassing the logic that used to skip it when we detected it was the same IP in use by more than one service.  Or if that's handled, can you help with what I missed?

Comment (champtar) [reply internal/layer2/announcer.go:229]: Good catch, indeed I spam once per name and not per IP, I'll change that

Comment (champtar) [reply internal/layer2/announcer.go]: Done, it even simplified a bit the code in the end

**Comment by champtar** [comment]
Note from my testing:
```
W1008 00:03:37.209528       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.ConfigMap ended with: too old resource version: 4356503 (4358042)
{"caller":"main.go:387","configmap":"metallb-system/config","event":"startUpdate","msg":"start of config update","ts":"2020-10-08T00:03:38.213385661Z"}
{"caller":"main.go:411","configmap":"metallb-system/config","event":"endUpdate","msg":"end of config update","ts":"2020-10-08T00:03:38.213495356Z"}
{"caller":"k8s.go:421","configmap":"metallb-system/config","event":"configLoaded","msg":"config (re)loaded","ts":"2020-10-08T00:03:38.213521216Z"}
{"caller":"main.go:268","event":"startUpdate","msg":"start of service update"
...
```
On each config watch expiry, we do a resync, so we spam

**Comment by russellb** [comment]
> Note from my testing:
> 
> ```
> W1008 00:03:37.209528       1 reflector.go:302] pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: watch of *v1.ConfigMap ended with: too old resource version: 4356503 (4358042)
> {"caller":"main.go:387","configmap":"metallb-system/config","event":"startUpdate","msg":"start of config update","ts":"2020-10-08T00:03:38.213385661Z"}
> {"caller":"main.go:411","configmap":"metallb-system/config","event":"endUpdate","msg":"end of config update","ts":"2020-10-08T00:03:38.213495356Z"}
> {"caller":"k8s.go:421","configmap":"metallb-system/config","event":"configLoaded","msg":"config (re)loaded","ts":"2020-10-08T00:03:38.213521216Z"}
> {"caller":"main.go:268","event":"startUpdate","msg":"start of service update"
> ...
> ```
> 
> On each config watch expiry, we do a resync, so we spam

how often does that happen?

File: internal/layer2/announcer.go
Comment (russellb) [inline internal/layer2/announcer.go:143]: It would be nice to turn this off when there's no active spamming happening.  Otherwise when it should be otherwise idle, it's still waking up every 1.1s for no reason.

Comment (champtar) [reply internal/layer2/announcer.go:143]: might be doable with ticker.Stop() and ticker.Reset(), will have a look

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:147]: I find `ips` confusing as it sounds like the var holds multiple IP addresses.

I suggest doing the following:

```go
case e := <-a.spamCh:
    ip := e.String()
```

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:25]:
```suggestion
	// This channel can block - do not write to it while holding the mutex
	// to avoid deadlocking.
```

- We should format the comments according to the existing style.
- Improve phrasing.

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:142]:
```suggestion
	// See https://github.com/metallb/metallb/issues/172 for the 1100 choice.
```

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:141]:
```suggestion
	// Map IP to spam stop time.
	m := map[string]time.Time{}
```

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:149]:
```suggestion
			// Set spam stop time to 5 seconds from now.
			m[ips] = time.Now().Add(5 * time.Second)
```

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:152]: - I don't understand this comment. Could you please make it clearer?

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:158]:
```suggestion
					// We have spammed enough - remove the IP from the map.
					delete(m, ips)
```

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:216]: How about the following?

```go
	defer func() {
		a.spamCh <- ip
	}()
```

We can then get rid of `spamStart()`. In any case, if we keep this function I'd rather name it `doSpam()` since having "start" in the name implies that there is also "stop", which isn't the case.

File: internal/layer2/announcer_test.go
Comment (johananl) [inline internal/layer2/announcer_test.go:31]:
```suggestion
		// We need to empty spamCh.
```

If you comment here I'd explain *why* we need to empty the channel.

File: internal/layer2/announcer.go
Comment (johananl) [inline internal/layer2/announcer.go:37]: What happens if the channel is full? What is the implication for the user?

**Review by johananl — COMMENTED** [review]
LGTM overall. The design looks OK to me and things seem to work on my testing.

Added some nitpicks and questions, mainly around comments.

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go:143]: ticker.Reset() is golang 1.15, do we want to bump minimum version to latest and greatest ? (even my Fedora only has 1.14)

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go]: merged

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go]: I went with `ipStr` because I also use `ip` in `a.spam(ip)`

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go]: Is it better ?
```
// The first 2 calls to spam() can be msec away,
// but we can't wait 1.1 sec
```

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go]: I went with `doSpam()`, just a matter of taste, but if other people prefer the anonymous func i'll change it

File: internal/layer2/announcer_test.go
Comment (champtar) [reply internal/layer2/announcer_test.go]: Changed:
// We need to empty spamCh as spamLoop() is not started.

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go:37]: If `spamCh` is full `SetBalancer` blocks until we are done spamming in `spamLoop` and empty `spamCh` again
I expect people to have less than 100 layer2 VIP per cluster so 1024 should be good. If I'm wrong full resync will block for a bit which is fine.

**Comment by champtar** [comment]
@johananl thanks a lot for the review, see my answers

**Comment by champtar** [comment]
BTW to speed up thing I'll implement an allow/deny list based on interface name, so we don't spam on each container interface when using calico or others

**Comment by russellb** [comment]
> BTW to speed up thing I'll implement an allow/deny list based on interface name, so we don't spam on each container interface when using calico or others. I've tried in the past to detect veth type but it's really not easy

that's not really related to this change though, right?

**Comment by champtar** [comment]
@russellb yes in a separate PR, at some point in the future

File: internal/layer2/announcer.go
Comment (johananl) [reply internal/layer2/announcer.go]: I still don't understand the technical meaning behind "but we can't wait 1.1 sec" :confused: What are we trying to say here?

Anyway, let's not block on this. I'd just change `msec` to `milliseconds` and add a period at the end for consistency.

**Review by johananl — APPROVED** [review]
LGTM.

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go]: If the ticker is already running (always the case right now), we might call spam() because it's a new IP, and 1msec after call spam() again in the ticker part. This is not ideal, but the worst case, where we wait 1099 ms for the first spam() call is not acceptable.

Comment (johananl) [reply internal/layer2/announcer.go]: I *think* I understand what you mean now. Can I rephrase the comment to the following?

```
				// Spam right away to avoid waiting 1100 milliseconds even if
				// it means we call spam() twice in a row in a short amount of
				// time.
```

Comment (champtar) [reply internal/layer2/announcer.go]: Done
```
// Spam right away to avoid waiting up to 1100 milliseconds even if
// it means we call spam() twice in a row in a short amount of time.
```

File: internal/layer2/announcer.go
Comment (champtar) [reply internal/layer2/announcer.go:144]: I'll open a issue as soon as we merge this PR so we don't forget

**Pull request merged**
